compiler/rustc_trait_selection/src/traits/select/mod.rs

```diff
@@ -1508,7 +1508,8 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
                 defining_opaque_types_and_generators: defining_opaque_types,
             }
             | TypingMode::Borrowck { defining_opaque_types } => {
-                defining_opaque_types.is_empty() || !pred.has_opaque_types()
+                defining_opaque_types.is_empty()
+                    || (!pred.has_opaque_types() && !pred.has_coroutines())
```

Comment:
i feel like we prolly want to instead never erase the coroutines, only the defining opaques when creating a `TypingEnv` from the current infcx

Reply:
might be expensive to erase on every call to `TypingEnv`; otherwise i'll keep this and write a comment explaining why it's unsound.